### ISSUE TITLE
v2.10.x: Raises `TypeError: mro() returned a non-class ('PydanticRecursiveRef')`

### ISSUE THREAD

**Issue by kim-youngseop-developer** [pydantic/pydantic#11024]
### Initial Checks

- [X] I confirm that I'm using Pydantic V2

### Description

After accidentally updating a package from 2.7.x to 2.10.x,
got an error message that I had never seen before.

I don't know if it's known issue or not,
but I'm just writing this in the hopes that it will help improve 2.10.x.
(Tested it and worked fine until 2.9.2)

Have a nice day! 👍

```
Traceback (most recent call last):
  File "/home/user-ubuntu/projects/~~~/.py", line 26, in <module>
    T12[int, int](values=[CombineModel[int, int](field_1=1, field_2=2)])
    ~~~^^^^^^^^^^
  File "/home/user-ubuntu/projects/~~~/.venv/lib/python3.11/site-packages/pydantic/main.py", line 811, in __class_getitem__
    submodel = _generics.create_generic_submodel(model_name, origin, args, params)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user-ubuntu/projects/~~~/.venv/lib/python3.11/site-packages/pydantic/_internal/_generics.py", line 137, in create_generic_submodel
    created_model = meta(
                    ^^^^^
  File "/home/user-ubuntu/projects/~~~/.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py", line 137, in __new__
    cls = cast('type[BaseModel]', super().__new__(mcs, cls_name, bases, namespace, **kwargs))
                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen abc>", line 106, in __new__
TypeError: mro() returned a non-class ('PydanticRecursiveRef')
```

### Example Code

```Python
from pydantic import BaseModel

from typing import List
from typing import TypeVar
from typing import Generic

_T = TypeVar("_T")

class GenericBaseModel(BaseModel, Generic[_T]):
    ...

# 😡 BAD CASE
# pydantic==2.10.2
# pydantic_core==2.27.1
# 😀 GOOD CASE
# pydantic==2.9.2
# pydantic_core==2.23.4
class EnumerableModel(GenericBaseModel[_T]):
    values: List[_T]

# 😀 GOOD CASE
# pydantic==2.10.2
# pydantic_core==2.27.1
# 😀 GOOD CASE
# pydantic==2.9.2
# pydantic_core==2.23.4
class EnumerableModel(BaseModel, Generic[_T]):
    values: List[_T]

_T1 = TypeVar("_T1")
_T2 = TypeVar("_T2")

class CombineModel(BaseModel, Generic[_T1, _T2]):
    field_1: _T1
    field_2: _T2

class EnumerableCombineModel(EnumerableModel[CombineModel[_T1, _T2]]):
    ...

EnumerableCombineModel[int, int]

# Traceback (most recent call last):
#   File "/home/user-ubuntu/projects/~~~/.py", line 29, in <module>
#     EnumerableCombineModel[int, int]
#     ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^
#   File "/home/user-ubuntu/projects/~~~/.venv/lib/python3.11/site-packages/pydantic/main.py", line 811, in __class_getitem__
#     submodel = _generics.create_generic_submodel(model_name, origin, args, params)
#                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#   File "/home/user-ubuntu/projects/~~~/.venv/lib/python3.11/site-packages/pydantic/_internal/_generics.py", line 137, in create_generic_submodel
#     created_model = meta(
#                     ^^^^^
#   File "/home/user-ubuntu/projects/~~~/.venv/lib/python3.11/site-packages/pydantic/_internal/_model_construction.py", line 137, in __new__
#     cls = cast('type[BaseModel]', super().__new__(mcs, cls_name, bases, namespace, **kwargs))
#                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
#   File "<frozen abc>", line 106, in __new__
# TypeError: mro() returned a non-class ('PydanticRecursiveRef')
```

### Python, Pydantic & OS Version

```Text
pydantic version: 2.10.2
        pydantic-core version: 2.27.1
          pydantic-core build: profile=release pgo=false
                 install path: /home/user-ubuntu/projects/metalicsoft-v1/.venv/lib/python3.11/site-packages/pydantic
               python version: 3.11.7 (main, Dec  8 2023, 18:56:58) [GCC 11.4.0]
                     platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.35
             related packages: fastapi-0.115.5 typing_extensions-4.12.2 pydantic-settings-2.6.1
                       commit: unknown
```

**Comment by sydney-runkle** [comment]
@kim-youngseop-developer,

Thanks for the report.

This is definitely something we want to fix in a v2.10.X patch.

**Comment by sydney-runkle** [comment]
cc @kc0506, I feel this might be related to the mro stuff...

**Comment by MarkusSintonen** [comment]
We are also having the same issue with 2.10.x versions. But we are seeing a different regression with the `mro()` where we are accessing the `mro`:
```
derived_from = some_generic_model.mro()
AttributeError: 'PydanticRecursiveRef' object has no attribute 'mro'
```

**Comment by sydney-runkle** [comment]
@MarkusSintonen,

Thanks for the report. Could you provide a full MRO for your use case? Thank you!

**Comment by MarkusSintonen** [comment]
@sydney-runkle :
```python
from typing import Generic, TypeVar
from pydantic import BaseModel
import pytest

@pytest.mark.parametrize("inner_mro_access", [True, False])  # Both fail but it used to work
def test_bug(inner_mro_access: bool):
    T = TypeVar("T")
    C = TypeVar("C")

    class BaseInner(BaseModel, Generic[T, C]):
        def __class_getitem__(cls, params):
            klass = super().__class_getitem__(params)
            if inner_mro_access:
                klass.mro()
            return klass

    class BaseOuter(BaseModel, Generic[T, C]):
        field1: BaseInner[T, C]

    class Inner(BaseInner[int, C], BaseModel, Generic[C]):
        pass

    class Outer(BaseOuter[int, C], BaseModel, Generic[C]):
        field2: Inner[C]

    # inner_mro_access=False. Fails: TypeError: mro() returned a non-class ('PydanticRecursiveRef')
    # inner_mro_access=True. Fails: AttributeError: 'PydanticRecursiveRef' object has no attribute 'mro'
    Outer[int]

```

**Comment by sydney-runkle** [comment]
Seems like this only shows up with pretty exotic use cases of generics. We certainly want to fix, but this will fall into the scope of v2.10.4
